mlflow/promptflow/__init__.py

```diff
@@ -0,0 +1,393 @@
+                       containing file dependencies). These files are *prepended* to the system
+                       path when the model is loaded.
+    :param mlflow_model: :py:mod:`mlflow.models.Model` this flavor is being added to.
+    :param signature: :py:class:`ModelSignature <mlflow.models.ModelSignature>`
+                      describes model input and output
+                      :py:class:`Schema <mlflow.types.Schema>`.
+                      If not specified, the model signature would be set according to
+                      `model.inputs` and `model.outputs` as columns names, `DataType.int`,
+                      `DataType.double`and `DataType.boolean` as the column type for
+                      promptflow type `int`, `double` and `bool`, and `DataType.string`
+                      as the column type for the other promptflow types.
+
+    :param input_example: {{ input_example }}
+    :param pip_requirements: {{ pip_requirements }}
+    :param extra_pip_requirements: {{ extra_pip_requirements }}
+    :param metadata: Custom metadata dictionary passed to the model and stored in the MLmodel file.
+
+                     .. Note:: Experimental: This parameter may change or be removed in a future
+                                             release without warning.
+    """
+    import promptflow
+    from promptflow import load_flow
+    from promptflow.contracts.tool import ValueType
+    from promptflow._sdk.entities._flow import Flow
+    from promptflow._sdk._utils import _merge_local_code_and_additional_includes
+    from promptflow._sdk.operations._run_submitter import remove_additional_includes
+
+    _validate_env_arguments(conda_env, pip_requirements, extra_pip_requirements)
+
+    # check if path exists
+    path = os.path.abspath(path)
+    _validate_and_prepare_target_save_path(path)
+
+    code_dir_subpath = _validate_and_copy_code_paths(code_paths, path)
+
+    model_flow_path = os.path.join(path, _MODEL_FLOW_DIRECTORY)
+
+    if not isinstance(model, Flow):
+        raise mlflow.MlflowException.invalid_parameter_value(
+            _UNSUPPORTED_MODEL_ERROR_MESSAGE.format(instance_type=type(model).__name__)
+        )
+
+    # Resolve additional includes in flow
+    with _merge_local_code_and_additional_includes(code_path=model.code) as resolved_model_dir:
+        remove_additional_includes(Path(resolved_model_dir))
+        shutil.copytree(src=resolved_model_dir, dst=model_flow_path)
+    # Get flow env in flow dag
+    flow_env = _resolve_env_from_flow(model.flow_dag_path)
+
+    def _parse_pf_type(type_str):
+        if type_str == ValueType.INT.value:
+            return DataType.integer
+        if type_str == ValueType.DOUBLE.value:
+            return DataType.double
+        if type_str == ValueType.BOOL.value:
+            return DataType.boolean
+        # Return string for list, object and other types
```

Comment:
is this the suggested way to fall back other types to string?

Reply:
There is no suitable type for list, object, image, etc. At the same time, I saw langchain dump all their inputs/outputs to string type.

Reply:
hey @brynn-code you might want to sync up with @serena-ruan  on her current work enabling arbitrary input types here: https://github.com/mlflow/mlflow/tree/llm_signature

this is probably the best alignment of your work, but please dig into the details there - at the end of the day the goal of serena's work is to enable stuff like what you're doing, so maybe we can check to see how aligned the design is

Reply:
@akshaya-a seems the arbitrary input type support has not finished yet, I would suggest we handle the signature in two steps:
1. continue this promptflow flavor contributing effort without signature in MLmodel
2. once that arbitrary type supporting is ready, we can update promptflow flavor to add back the signature.

We'd better not introduce hard dependencies between these two efforts.

Reply:
llm_signature is planned to release with 2.10.0, and ideally you should be able to reuse whatever we supports within infer_signature directly :)

Reply:
Offline synced and add _infer_signature_from_input_example.